### PR TITLE
[phrase] range searches for QueryPhrases with terminal prefixes

### PR DESCRIPTION
**Pull request by boblannon**
This PR adds an additional bit of logic to `PhraseSet::contains_prefix`, allowing the final `QueryWord` in a `QueryPhrase` to be a `QueryWord::Prefix`.  When that's the case, it will do a search over the range of possible word ids.

### CODE REVIEW

File: src/phrase/mod.rs [hunk 1]
```diff
@@ -70,6 +73,222 @@ impl PhraseSet {
...
+    fn contains_prefix_with_range(&self, phrase: QueryPhrase) -> bool {
+        let (sought_min_key, sought_max_key) = phrase.prefix_key_range().unwrap();
+
+		// self as fst
+        let fst = &self.0.as_fst();
+        // start from root node
+        let root_node = fst.root();
+
+		// using the keys for the full words, walk the graph. if no path accepts these keys, stop
+        // here. result node should not be final.
```
Comment (apendleton) [inline src/phrase/mod.rs]: It can be final as long as it's not exclusively so, right? (this is what your code does, I think, but might be good to clarify the comment)

File: src/phrase/mod.rs [hunk 2]
```diff
@@ -70,6 +73,222 @@ impl PhraseSet {
...
+            Some(..) => { return true },
+            _ => (),
+        }
+
+        // does the key at the high end of the prefix range take us to a final state? if so, we know
+        // that at least one of the possible phrases is in the graph
+        match self.partial_search(full_word_addr, &sought_max_key) {
+            Some(..) => { return true },
+            _ => (),
+        }
```
Comment (apendleton) [inline src/phrase/mod.rs]: This pair of checks (looking for the sought min and sought max) feels pretty unproductive to me. The sought min and sought max are going to be the IDs of the very first word with the given prefix, lexicographically, of the whole corpus, and likewise the last. So if you've typed "main s", the prefix for S is going to have the IDs of, like, "Saarinen" and "Szechuan," or whatever. The odds that those are the next word we're actually going to find are tiny. Further, the situations where either of these checks would be true are one of three situations: when the actual and sought min and max are interleaved (in one direction or the other), or when the actual range entirely encompasses the sought range -- but most of the time when we're in one of these situations, these checks will still fail, even if ultimately we're going to return true, because we don't happen to have Saarinen or Szechuan specifically. We're checking for those cases anyway, below, and their framed around checking actual exemplars meaning if we're in that state, we'll always actually catch it, rather than taking a shot in the dark here and probably failing.

File: src/phrase/mod.rs [hunk 3]
```diff
@@ -70,6 +73,222 @@ impl PhraseSet {
...
+                    match nl {
+                        Look::Stop => return true,
+                        _ => {
+                            next_looks.push(nl);
+                        },
+                    }
+                }
+            }
+
+            looks = next_looks.drain(..).collect::<Vec<Look>>();
```
[the remark this answers is not in the document]
Comment (boblannon) [reply src/phrase/mod.rs]: ah yeah, okay, this is vestigial from a previous iteration, where `next_looks` wasn't re-init'd every loop.  either way, though, i think `append` is the right call.

[the remark this answers is not in the document]
Comment (boblannon) [reply src/phrase/mod.rs]: ....actually it looks like i need to shadow `looks` because it's moved in the in the `into_iter` loop above.  So i need to derive `Copy` and `Clone` on the `Look` enum, and clone.

```rust
looks = next_looks.clone()
```

[the remark this answers is not in the document]
Comment (apendleton) [reply src/phrase/mod.rs]: Ah that's unfortunate. Boo unnecessary copies. What if you changed the `into_iter` to `drain` (so, consume the contents of the vector, but not the vector itself)? Or alternatively, if you made range_search take a reference to a look rather than moving it, and then used a regular `iter` instead of `into_iter`.

File: src/phrase/mod.rs (hunk 2 above)
Comment (boblannon) [reply src/phrase/mod.rs]: ah, yes, good point.  the earlier iteration relied on the edge cases being ruled out, but these are now redundant with the subsequent overlap tests.

File: src/phrase/mod.rs (hunk 1 above)
Comment (boblannon) [reply src/phrase/mod.rs]: yep, i removed the checks for finalness but didn't update the doc comment.

File: src/phrase/mod.rs [hunk 4]
```diff
@@ -70,6 +73,209 @@ impl PhraseSet {
...
+                    match nl {
+                        Look::Stop => return true,
+                        _ => {
+                            next_looks.push(nl);
+                        },
+                    }
+                }
+            }
+
+            looks = next_looks;
```
Comment (boblannon) [inline src/phrase/mod.rs]: @apendleton the solution was just to re-assign, which moves the vector

Comment (apendleton) [reply src/phrase/mod.rs]: Right yeah, this creates a new vector at the top of each loop (the new `new_looks`) and destroys one at the bottom (the old `looks`); the `append` solution was an attempt at just reusing the same two to avoid the new allocation. This works for me though.

Comment (boblannon) [reply src/phrase/mod.rs]: yeah, append didn't work because the `looks` vector was moved in the loop above.  if there's another way of doing that loop so that it doesn't, then append would be perfect.

**Comment by boblannon** [comment]
@apendleton I was actually thinking that I should probably include some extra documentation to explain the thought process here.  i think it'd be helpful to include some images to do that, but that's not supported in rustdoc yet.  does it make sense to just add a README.md to the module folder?
